lib/content_script/init.js

```diff
@@ -0,0 +1,41 @@
+'use strict'
+
+/* global nodeProcess, isolatedWorld */
+
+const events = require('events')
+const { EventEmitter } = events
+
+process.atomBinding = require('@electron/internal/common/atom-binding-setup')(nodeProcess.binding, 'renderer')
+
+const v8Util = process.atomBinding('v8_util')
+// The `lib/renderer/ipc-renderer-internal.js` module looks for the ipc object in the
+// "ipc-internal" hidden value
+v8Util.setHiddenValue(global, 'ipc-internal', new EventEmitter())
+// The process object created by browserify is not an event emitter, fix it so
+// the API is more compatible with non-sandboxed renderers.
+for (const prop of Object.keys(EventEmitter.prototype)) {
+  if (process.hasOwnProperty(prop)) {
+    delete process[prop]
+  }
+}
+Object.setPrototypeOf(process, EventEmitter.prototype)
+
+const isolatedWorldArgs = v8Util.getHiddenValue(isolatedWorld, 'isolated-world-args')
+
+if (isolatedWorldArgs) {
+  const { ipcRenderer, guestInstanceId, isHiddenPage, openerId, usesNativeWindowOpen } = isolatedWorldArgs
+  require('@electron/internal/renderer/window-setup')(ipcRenderer, guestInstanceId, openerId, isHiddenPage, usesNativeWindowOpen)
+}
+
+const chromeAPI = require('@electron/internal/renderer/chrome-api')
+
+// Await initialization with extension ID
+window.__init = (extensionId) => {
```

Comment:
I feel as though there should be a better way of doing this, it also implies that this initalizer and the users chrome extension aren't isolated?

I may be missing where the isolation boundary is that this PR is trying to create.

E.g.

`Chrome Extension |||| Electron Code` but this PR appears to be more `Chrome Extension + Electron Code ||| User Code` which implies a high level of trust in the Chrome Extension code?  Maybe I just need some clarification on that

Reply:
The `__init` function is coupled with Electron code in the `content-scripts-injector.ts` script. Prior to running a Chrome extension's content script, it will call the `__init` function with the associated `extensionId`.
```ts
const runContentScript = function (this: any, extensionId: string, url: string, code: string) {
  const sources = [
    // initialize Chrome API in isolated world
    { code: `typeof __init === 'function' && __init('${extensionId}')` },
    // then run content script in isolated world
    { code, url }
  ]

  // ...

  webFrame.executeJavaScriptInIsolatedWorld(worldId, sources)
}
```
Since `__init` is called first from Electron code, user code won't be able to access it after it has been removed. The exception would be if the user creates an isolated world within the reserved extension ID range, outside of the content script injector.

This still feels dirty to me, but I'm not yet sure how to initialize the `content_script_bundle` with the extension ID ahead of time.

Reply:
@MarshallOfSound I found another way to pass the extension ID to the `content_script_bundle` script using a v8 private. fb42cd0345d0efb47a994c5d2444253f89dcd5d1

Now there is a clear isolation between user code and Electron code for initializing the content script isolated world.